pdcupdater/handlers/modules.py

```diff
@@ -0,0 +1,267 @@
+import os
+import logging
+import errno
+import re
+from subprocess import check_call, STDOUT
+
+import beanbag
+import pdcupdater.handlers
+import pdcupdater.services
+from pdcupdater.utils import TmpDir, PushPopD
+
+import modulemd
+
+log = logging.getLogger(__name__)
+
+
+class ModuleStateChangeHandler(pdcupdater.handlers.BaseHandler):
+    """ When the state of a module changes. """
+
+    relevant_states = ('done', 'ready')
+    valid_states = relevant_states + ('init', 'wait', 'failed')
+
+    tree_id_re = re.compile(
+        r"(?P<name>[^-]+)-(?P<version>[^-]+)-"
+        r"(?P<date>[0-9]{8})\.(?P<spin>[0-9]+)")
+
+    rpm_fname_re = re.compile(
+        r"(?P<name>.+)-"
+        r"(?:(?P<epoch>[^:]+):)?(?P<version>[^-:]+)-(?P<release>[^-]+)."
+        r"(?P<arch>[^\.]+).rpm")
+
+    scmurl_re = re.compile(
+        r"(?P<giturl>(?:(?P<scheme>git)://(?P<host>[^/]+))?"
+        r"(?P<repopath>/[^\?]+))\?(?P<modpath>[^#]*)#(?P<revision>.+)")
+
+    @property
+    def topic_suffixes(self):
+        return ['module.state.change']
+
+    def can_handle(self, msg):
+        if not msg['topic'].endswith('module.state.change'):
+            return False
+
+        state = msg['msg']['state']
+
+        if state not in self.valid_states:
+            log.warn("Invalid module state '{}', skipping.".format(state))
+            return False
+
+        if state not in self.relevant_states:
+            log.debug("Non-relevant module state '{}', skipping.".format(
+                state))
+            return False
+
+        return True
+
+    def handle(self, pdc, msg):
+        body = msg['msg']
+        state = body['state']
+
+        if state not in self.relevant_states:
+            log.warn("Non-relevant module state '{}', skipping.".format(
+                state))
+            return
+
+        unreleased_variant = self.get_or_create_unreleased_variant(pdc, body)
+
+        if 'topdir' in body:
+            self.handle_new_tree(pdc, body, unreleased_variant)
+
+    def get_mmd_from_scm(self, scmurl):
+        with TmpDir(prefix="pdcupdater-") as tmpdir, PushPopD(tmpdir), \
+                open(os.devnull, "w") as devnull:
+            m = self.scmurl_re.match(scmurl)
+            if not m:
+                log.error("Can't parse SCM URL: {}".format(scmurl))
+                raise RuntimeError("Can't parse SCM URL: {}".format(scmurl))
+            giturl = m.group('giturl')
+            repopath = m.group('repopath')
+            modpath = m.group('modpath')
+            revision = m.group('revision')
+            while repopath.endswith("/"):
+                repopath = repopath[:-1]
+            modname = repopath.rsplit("/", 1)[1]
+            if modname.endswith(".git"):
+                modname = modname[:-4]
+
+            log.debug("Cloning {}".format(giturl))
+            check_call(["git", "clone", "-n", giturl, modname],
```

Comment:
I'm indifferent.  If it raises an exception, that will get caught and logged.

Does the CalledProcessError have information about the command that it tried to run?  As long as it has that, that will be enough information to debug production issues in the future.